### PR TITLE
Pn 7211 api key paginazione assente

### PR DESCRIPTION
**Pull request by ImJimMoriarty**
## Short description
Added CustomPagination then changed the action and reducer so that the pagination can process the right page requested.

## List of changes proposed in this pull request

- Added CustomPagination component.
- Changed Api request for the ApiKeys so that it can retrieve the full object instead of the array of keys.
- Changed GetApiKeysResponse model by adding the attribute "total" so that it can be used into CustomPagination.
- Added 2 new models: ApiKeyParam, ApiKeyFull.
- Changed actions by adding ApiKeyParam and returning an ApiKeyFull to manage all the pagination attributes.
- Added new reducer "setPagination" to manage the pagination logic.
- Changed initialState by adding new attributes.
- Added new events to track the ApiKeys's table pagination.

## How to test
Just go into the API Key section of PA and try to change page or the size of the table.

### CODE REVIEW

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx [hunk 1]
```diff
@@ -85,14 +95,32 @@ const ApiKeys = () => {
   const { t } = useTranslation(['apikeys']);
 
   const apiKeys = useAppSelector((state: RootState) => state.apiKeysState.apiKeys);
+  const pagination = useAppSelector((state: RootState) => state.apiKeysState.pagination);
+
+  const totalElements =
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx]: This is the pagination management for a differente structure. In this case we don't have moreResult or nextPagesKey.

So 
- totalElement = total
- pagesToShow = Math.min(pages, 3) where pages = Math.ceil(total / size)

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx [hunk 2]
```diff
@@ -85,14 +95,32 @@ const ApiKeys = () => {
...
+    Math.min(pagination.nextPagesKey.length + 1, 3),
+    pagination.page + 1
+  );
 
   const fetchApiKeys = useCallback(() => {
-    void dispatch(getApiKeys());
-  }, []);
+    const params = {
+      limit: pagination.size,
+      lastKey: pagination.page === 0 ? undefined : pagination.nextPagesKey[pagination.page - 1],
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx:111]: In the redux slice, store lastKey and lastUpdate. See the comment below, in the reducer file

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx [hunk 3]
```diff
@@ -139,6 +167,16 @@ const ApiKeys = () => {
     void dispatch(deleteApiKey(apiKeyId)).then(() => void dispatch(getApiKeys()));
   };
 
+  // Pagination handlers
+  const handleChangePage = (paginationData: PaginationData) => {
+    trackEventByType(TrackEventType.NOTIFICATION_TABLE_PAGINATION);
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx:168]: this track event is wrong. this function logs on a platform called mixpanel and with this event we are logging that an event of type "notification pagination" occurred

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx [hunk 4]
```diff
@@ -139,6 +167,16 @@ const ApiKeys = () => {
...
+    dispatch(setPagination({ size: paginationData.size, page: paginationData.page }));
+  };
+
+  const handleEventTrackingCallbackPageSize = (pageSize: number) => {
+    trackEventByType(TrackEventType.NOTIFICATION_TABLE_SIZE, { pageSize });
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx:173]: same as before

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts [hunk 5]
```diff
@@ -1,10 +1,21 @@
...
 import { getApiKeys } from './actions';
 
 const initialState = {
   loading: false,
-  apiKeys: [] as Array<ApiKey>,
+  apiKeys: {
+    items: [],
+    total: 0,
+  } as ApiKeys<UserGroup>,
+  pagination: {
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts:12]: in pagination we can store the size (= limit), the page and an array of object like that

`{lastKey: ..., lastUpdate: ...}`

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts [hunk 6]
```diff
@@ -13,14 +24,28 @@ const apiKeysSlice = createSlice({
   initialState,
   reducers: {
     resetState: () => initialState,
+    setPagination: (state, action: PayloadAction<{ page: number; size: number }>) => {
+      if (state.pagination.size !== action.payload.size) {
+        // reset pagination
+        state.pagination.nextPagesKey = [];
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts:30]: here we will reset only the array of object `{lastKey: ..., lastUpdate: ...}`

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts [hunk 7]
```diff
@@ -13,14 +24,28 @@ const apiKeysSlice = createSlice({
...
+      }
+      state.pagination.size = action.payload.size;
+      state.pagination.page = action.payload.page;
+    },
   },
   extraReducers: (builder) => {
     builder.addCase(getApiKeys.fulfilled, (state, action) => {
       state.apiKeys = action.payload;
+      if (action.payload.lastKey && action.payload.lastUpdate) {
+        state.pagination.nextPagesKey.push(action.payload.lastKey);
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts:42]: here we push into the array of object `{lastKey: ..., lastUpdate: ...}`. pay attention to not push the same object twice

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx (hunk 1 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx]: https://github.com/pagopa/pn-frontend/pull/926/commits/c7ee0db13db237837b18444852479d492db22fc6

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx (hunk 3 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx:168]: Yeah i actually forgot to replace them even though i actullay added those 2 new type of trackEvent.

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx (hunk 2 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx (hunk 3 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts (hunk 6 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts:32]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts (hunk 7 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

File: packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts (hunk 5 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/redux/apiKeys/reducers.ts:12]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

File: packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx (hunk 4 above)
Comment (ImJimMoriarty) [reply packages/pn-pa-webapp/src/pages/ApiKeys.page.tsx]: changes made into 5eb41f24bacd7862f6c9ac99ae9f9e9396e71a40

**Pull request merged**